reports/module_cost_variation.go

```diff
@@ -0,0 +1,230 @@
+	SheetName:     costVariationLast6MonthsSheetName,
+	ErrorName:     "costVariationLastSixMonthsError",
+	GenerateSheet: costVariationGenerateLast6Months,
+}
+
+func costVariationGenerateLastMonth(ctx context.Context, aas []aws.AwsAccount, date time.Time, _ *sql.Tx, file *excelize.File) (err error) {
+	var dateRange diff.DateRange
+	if date.IsZero() {
+		dateRange.Begin, dateRange.End = history.GetHistoryDate()
+	} else {
+		dateRange = diff.DateRange{
+			Begin: date,
+			End:   time.Date(date.Year(), date.Month()+1, 0, 23, 59, 59, 999999999, date.Location()).UTC(),
+		}
+	}
+	dates := make([]time.Time, dateRange.End.Day())
+	for index := range dates {
+		dates[index] = dateRange.Begin.AddDate(0, 0, index)
+	}
+	frequency := costVariationFrequency{costVariationLastMonthSheetName, "Daily Cost", "day", "2006-01-02", dates}
+	return costVariationGenerateSheet(ctx, aas, dateRange, frequency, file)
+}
+
+func costVariationGenerateLast6Months(ctx context.Context, aas []aws.AwsAccount, date time.Time, _ *sql.Tx, file *excelize.File) (err error) {
+	var dateRange diff.DateRange
+	if date.IsZero() {
+		_, dateRange.End = history.GetHistoryDate()
+	} else {
+		dateRange.End = time.Date(date.Year(), date.Month()+1, 0, 23, 59, 59, 999999999, date.Location()).UTC()
+	}
+	dateRange.Begin = time.Date(dateRange.End.Year(), dateRange.End.Month()-5, 1, 0, 0, 0, 0, dateRange.End.Location()).UTC()
+	dates := make([]time.Time, 6)
+	for index := range dates {
+		dates[index] = dateRange.Begin.AddDate(0, index, 0)
+	}
+	frequency := costVariationFrequency{costVariationLast6MonthsSheetName, "Monthly Cost", "month", "2006-01", dates}
+	return costVariationGenerateSheet(ctx, aas, dateRange, frequency, file)
+}
+
+func costVariationGenerateSheet(ctx context.Context, aas []aws.AwsAccount, dateRange diff.DateRange, frequency costVariationFrequency, file *excelize.File) (err error) {
+	data, err := costVariationGetData(ctx, aas, dateRange, frequency)
+	if err == nil {
+		return costVariationInsertDataInSheet(ctx, aas, dateRange, frequency, file, data)
+	} else {
+		return
+	}
+}
+
+func costVariationGetData(ctx context.Context, aas []aws.AwsAccount, dateRange diff.DateRange, frequency costVariationFrequency) (data map[aws.AwsAccount]costVariationReport, err error) {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	logger.Debug("Getting Cost Variation Report for accounts", map[string]interface{}{
+		"accounts":    aas,
+		"dateStart":   dateRange.Begin,
+		"dateEnd":     dateRange.End,
+		"aggregation": frequency.Aggregation,
+	})
+	data = make(map[aws.AwsAccount]costVariationReport, len(aas))
+	for _, account := range aas {
+		report, err := diff.TaskDiffData(ctx, account, dateRange, frequency.Aggregation)
+		if err != nil {
+			logger.Error("An error occurred while generating a Cost Variation Report", map[string]interface{}{
+				"error":     err,
+				"account":   account,
+				"dateStart": dateRange.Begin,
+				"dateEnd":   dateRange.End,
+			})
+			return data, err
```

Comment:
I call this function in `costVariationGenerateSheet` and it will check if `err` is `nil` or not. And also, in the next function where I use `data`, I use `range` to go through it.